### PR TITLE
feat(decorators): add three new decorators

### PR DESCRIPTION
**Pull request by JesseTG**
RepeatForever, RepeatUntilSuccess, and RepeatUntilFailure are new decorators used to continue a task
indefinitely.  RepeatForever executes its child and returns TaskStatus.Continue, regardless of its
child's return status.  RepeatUntilSuccess returns TaskStatus.Success if its child returns success,
else it returns TaskStatus.Continue.  RepeatUntilFailure returns TaskStatus.Failure if its child
returns failure, else it returns TaskStatus.Continue.

### CODE REVIEW

**Comment by ashblue** [comment]
[Until success decorator](https://trello.com/c/ayLl96lw/17-until-success-decorator)

**Comment by ashblue** [comment]
[Repeat decorator](https://trello.com/c/ODdAEC2W/15-repeat-decorator)

**Comment by ashblue** [comment]
[Until failure decorator](https://trello.com/c/ILtJACZe/16-until-failure-decorator)

**Comment by ashblue** [comment]
@JesseTG Just realizing there aren't any docs. My bad for not listing that anywhere (put a task in this week to setup some contribution guidelines that specify a checklist for PRs). If you have some spare time to throw in some `README.md` docs for each action that'd be awesome. If not no worries and I should be able to throw in the docs this weekend.

**Comment by JesseTG** [comment]
Oops.  My bad.  Yeah, I'll whip them up today or tomorrow.

File: Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilFailure.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public class RepeatUntilFailure : DecoratorBase {
+        protected override TaskStatus OnUpdate () {
+            if (Child == null) {
+                return TaskStatus.Failure;
+            }
+
+            var childStatus = Child.Update();
+            var status = childStatus;
+
+            if (status == TaskStatus.Failure) {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilFailure.cs]: Just run `Child.Update()` here

File: Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilFailure.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+                return TaskStatus.Failure;
+            }
+            else {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilFailure.cs]: No reason for else wrapper, remove.

File: Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilSuccess.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+using CleverCrow.Fluid.BTs.Trees;
+using CleverCrow.Fluid.BTs.Decorators;
+using CleverCrow.Fluid.BTs.Tasks;
+
+namespace CleverCrow.Fluid.BTs.Decorators {
+    public class RepeatUntilSuccess : DecoratorBase {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilSuccess.cs]: Same adjustments as `RepeatUntilFailure` class

File: Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatForeverTest.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+using CleverCrow.Fluid.BTs.Decorators;
+using CleverCrow.Fluid.BTs.Tasks;
+using NUnit.Framework;
+
+namespace CleverCrow.Fluid.BTs.Testing {
+    public class RepeatForeverTest {
+        public class UpdateMethod {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatForeverTest.cs]: Move repeated logic into a `[Setup]` method. Mostly just the `repeater` declaration and creation.

File: Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilFailureTest.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+using CleverCrow.Fluid.BTs.Decorators;
+using CleverCrow.Fluid.BTs.Tasks;
+using NUnit.Framework;
+
+namespace CleverCrow.Fluid.BTs.Testing {
+    public class RepeatUntilFailureTest {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilFailureTest.cs]: Move repeated logic into a [Setup] method. Mostly just the repeater declaration and creation.

File: Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilFailureTest.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+            [Test]
+            public void Returns_continue_on_child_continue () {
+                var repeater = new RepeatUntilFailure();
+                repeater.AddChild(A.TaskStub().WithUpdateStatus(TaskStatus.Continue).Build());
+
+                Assert.AreEqual(TaskStatus.Continue, repeater.Update());
+            }
+
+            [Test]
+            public void Does_not_crash_if_no_child () {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilFailureTest.cs]: Remove in favor of putting this on the base class

File: Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilSuccessTest.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
+using CleverCrow.Fluid.BTs.Decorators;
+using CleverCrow.Fluid.BTs.Tasks;
+using NUnit.Framework;
+
+namespace CleverCrow.Fluid.BTs.Testing {
+    public class RepeatUntilSuccessTest {
+        public class UpdateMethod {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilSuccessTest.cs]: Move repeated logic to a `[Setup]` method

File: Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilSuccessTest.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+            [Test]
+            public void Returns_continue_on_child_continue () {
+                var repeater = new RepeatUntilSuccess();
+                repeater.AddChild(A.TaskStub().WithUpdateStatus(TaskStatus.Continue).Build());
+
+                Assert.AreEqual(TaskStatus.Continue, repeater.Update());
+            }
+
+            [Test]
+            public void Does_not_crash_if_no_child () {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatUntilSuccessTest.cs]: Remove in favor of moving up to the base class.

File: Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilFailure.cs [hunk 9]
```diff
@@ -0,0 +1,23 @@
+using CleverCrow.Fluid.BTs.Trees;
+using CleverCrow.Fluid.BTs.Decorators;
+using CleverCrow.Fluid.BTs.Tasks;
+
+namespace CleverCrow.Fluid.BTs.Decorators {
+    public class RepeatUntilFailure : DecoratorBase {
+        protected override TaskStatus OnUpdate () {
+            if (Child == null) {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilFailure.cs]: I really like this. I'm starting to think that returning failure on `Child == null` should be the case for all decorators. This should probably be moved up to the `DecoratorBase` and remove any missing child checks on decorators / test. Much cleaner approach.

File: Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilSuccess.cs [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+                return TaskStatus.Success;
+            }
+
+            var childStatus = Child.Update();
+            var status = childStatus;
+
+            if (status == TaskStatus.Success) {
+                return TaskStatus.Success;
+            }
+            else {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilSuccess.cs]: Remove else wrapper

File: Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilSuccess.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    public class RepeatUntilSuccess : DecoratorBase {
+        protected override TaskStatus OnUpdate () {
+            if (Child == null) {
+                return TaskStatus.Success;
+            }
+
+            var childStatus = Child.Update();
+            var status = childStatus;
+
+            if (status == TaskStatus.Success) {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/Decorators/RepeatUntilSuccess.cs]: Move `Child.Update()` into if

File: Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatForeverTest.cs [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            [Test]
+            public void Returns_continue_on_child_continue () {
+                var repeater = new RepeatForever();
+                repeater.AddChild(A.TaskStub().WithUpdateStatus(TaskStatus.Continue).Build());
+
+                Assert.AreEqual(TaskStatus.Continue, repeater.Update());
+            }
+
+            [Test]
+            public void Does_not_crash_if_no_child () {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Tests/Editor/Decorators/RepeatForeverTest.cs]: Remove in favor of moving this up to the base class.

File: Assets/FluidBehaviorTree/Runtime/BehaviorTree/Builder/BehaviorTreeBuilder.cs [hunk 13]
```diff
@@ -62,6 +62,18 @@ public class BehaviorTreeBuilder {
             return ParentTask<ReturnFailure>(name);
         }
         
+        public BehaviorTreeBuilder RepeatUntilSuccess (string name = "repeat until success") {
```
Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/BehaviorTree/Builder/BehaviorTreeBuilder.cs]: Add docs to `README.md` for all 3 decorator types and update the table of contents.

Comment (ashblue) [inline Assets/FluidBehaviorTree/Runtime/BehaviorTree/Builder/BehaviorTreeBuilder.cs]: Also adding an icon is suggested for the new tree visualizer. Currently using Material Icons. One like this should work for all 3 https://material.io/tools/icons/?search=repeat&icon=repeat&style=baseline

Comment (ashblue) [reply Assets/FluidBehaviorTree/Runtime/BehaviorTree/Builder/BehaviorTreeBuilder.cs]: PS You'll need the latest `develop` for icons.

**Comment by ashblue** [comment]
I've pushed a commit that will remove the need for any missing child decorator checks. Pull the latest and it should allow you to remove the corresponding logic as noted in the comments.

**Comment by ashblue** [comment]
@all-contributors please add @JesseTG for code

**Pull request merged**